core/src/main/java/org/elasticsearch/common/ParseFieldMatcher.java

```diff
@@ -56,4 +61,35 @@ public boolean isStrict() {
     public boolean match(String fieldName, ParseField parseField) {
         return parseField.match(fieldName, strict);
     }
+
+    /**
+     * Select a named thing from a registry of named things while checking that the name matches the ParseField in the context of this
+     * matcher. Used for looking up parsers for queries, score functions, and aggregations.
+     *
+     * @param registryOf The kind of things stored in the registry. Used to make error messages. Should be singular like "query" instead of
+     *        "queries".
+     * @param registry The actual registry.
+     * @param name The name of the thing to look up.
+     * @param xContentLocation The location in the XContent from which the name was looked up. This is used for error message.
+     * @param nullOnNotFound If true then return null rather than throw an exception when the entry isn't found.
+     * @return The thing being looked up. Never null if nullOk is false.
+     * @throws ParsingException if the named thing isn't in the registry or the name was deprecated and deprecated names aren't supported.
+     */
+    public <T> T selectFromRegistry(String registryOf, Map<String, Tuple<ParseField, T>> registry, String name,
```

Comment:
Could we instead make the registry its own class that you can construct (the constructor would take the `registryOf` `nullOnNotFound` and `registry` values) and then we call the get method with the `name`, `xContentLocation` and `parseFieldMatcher`. This would mean that the code is abstracted into a common place without trying to shoehorn it into ParserFieldMatcher.